x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java

```diff
@@ -335,16 +336,20 @@ void run() throws IOException {
                  * So! We take the least common denominator which is the loader
                  * from the element expected element type.
                  */
-                builders[f] = fields[f].info.type.newBlockBuilder(docs.getPositionCount(), blockFactory);
+                fieldTypeBuilders[f] = fields[f].info.type.newBlockBuilder(docs.getPositionCount(), blockFactory);
+                builders[f] = new Block.Builder[shardContexts.size()];
+                converters[f] = new BlockLoader[shardContexts.size()];
```

Comment:
I remember originally trying the combine these, but seem to remember it causing hassles with the ColumnAtATimeReader. This was before I really got the row-stride reader working, so perhaps it is time for another attempt.

Reply:
I'd be ok delaying it to a followup too.

Reply:
I realise a key difference between what you suggested and what I tried (and was thinking of trying again). You said to put the builder and converter inside the loader. Instead I had tried to put the converter inside the builder. This is because we will always have both a builder and loader (the builder is made by the loader once additional information is known using code like `field.loader.builder(loaderBlockFactory, docs.count())`), so merging the loader and builder seems unworkable. Right now the converter is the only thing at play. It is currently implemented by the loader, since it does not need any of the information in the builder. But this does lead to lines like this:

```
Block build() {
    return (Block) loader.convert(builder.build());
}
```

If I put the converter into the builder, we can hide it inside the `build()` method call, simplifying things slightly. That was what I was trying before, and could try again.